src/Sarif.Multitool/Rules/SARIF1022.PreferMessageIdInsteadOfMessageText.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.CodeAnalysis.Sarif.Multitool.Rules
+{
+    public class PreferMessageIdInsteadOfMessageText : SarifValidationSkimmerBase
+    {
+        private IList<ReportingDescriptor> currentRules;
+
+        private readonly MultiformatMessageString _fullDescription = new MultiformatMessageString
+        {
+            Text = RuleResources.SARIF1022_PreferMessageIdInsteadOfMessageText
+        };
+
+        public override MultiformatMessageString FullDescription => _fullDescription;
+
+        public override FailureLevel DefaultLevel => FailureLevel.Error;
+
+        /// <summary>
+        /// SARIF1019
+        /// </summary>
+        public override string Id => RuleId.PreferMessageIdInsteadOfMessageText;
+
+        protected override IEnumerable<string> MessageResourceNames => new string[]
+        {
+            nameof(RuleResources.SARIF1022_RuleMetadataDoesntIncludeMessageId),
+            nameof(RuleResources.SARIF1022_RuleMetadataIsNotConsistentWithArguments),
+            nameof(RuleResources.SARIF1022_ResultShouldUseMessageIdInsteadOfMessageText)
+        };
+
+        protected override void Analyze(Run run, string runPointer)
+        {
+            this.currentRules = run?.Tool?.Driver?.Rules;
+        }
+
+        protected override void Analyze(Result result, string resultPointer)
+        {
+            // if message.text is not null, we have a problem
+            if (!string.IsNullOrEmpty(result.Message.Text))
+            {
+                LogResult(resultPointer, nameof(RuleResources.SARIF1022_ResultShouldUseMessageIdInsteadOfMessageText), result.Message.Text);
+                return;
+            }
+
+            // if message.id is not null, we have to check if that exists in rules
+            if (!string.IsNullOrEmpty(result.Message.Id))
+            {
+                // we dont have rules or we didnt find a match, we have a problem
+                if (this.currentRules == null || this.currentRules.Count == 0 || this.currentRules.All(r => !r.MessageStrings.ContainsKey(result.Message.Id)))
```

Comment:
>ContainsKey [](start = 126, length = 11)

BUG: We don't want to look in all the rules. We just want to look in the rule with the `id` that matches the `ruleId` property (or the `rule.id` property) of _this_ result.

Reply:
the new condition:
```csharp
if (this.currentRules == null
    || this.currentRules.Count == 0
    || (this.currentRules.Any() && !this.currentRules.FirstOrDefault(r => r.Id == (result.RuleId ?? result.Rule.Id)).MessageStrings.ContainsKey(result.Message.Id)))
{
    LogResult(resultPointer, nameof(RuleResources.SARIF1022_ResultMessageUsesTextProperty), result.Message.Id, (result.RuleId ?? result.Rule.Id));
    return;
}
```